presto-kafka/src/main/java/io/prestosql/plugin/kafka/KafkaPageSink.java

```diff
@@ -0,0 +1,156 @@
+import io.prestosql.spi.type.TimestampWithTimeZoneType;
+import io.prestosql.spi.type.Type;
+import io.prestosql.spi.type.VarcharType;
+import org.apache.kafka.clients.producer.KafkaProducer;
+import org.apache.kafka.clients.producer.ProducerRecord;
+import org.apache.kafka.common.serialization.ByteArraySerializer;
+
+import java.util.Collection;
+import java.util.List;
+import java.util.concurrent.CompletableFuture;
+
+import static io.prestosql.spi.type.BigintType.BIGINT;
+import static io.prestosql.spi.type.BooleanType.BOOLEAN;
+import static io.prestosql.spi.type.DateType.DATE;
+import static io.prestosql.spi.type.DoubleType.DOUBLE;
+import static io.prestosql.spi.type.IntegerType.INTEGER;
+import static io.prestosql.spi.type.RealType.REAL;
+import static io.prestosql.spi.type.SmallintType.SMALLINT;
+import static io.prestosql.spi.type.TimeType.TIME;
+import static io.prestosql.spi.type.TimeWithTimeZoneType.TIME_WITH_TIME_ZONE;
+import static io.prestosql.spi.type.TinyintType.TINYINT;
+import static io.prestosql.spi.type.VarbinaryType.VARBINARY;
+import static java.lang.Float.intBitsToFloat;
+import static java.lang.Math.toIntExact;
+import static java.util.Objects.requireNonNull;
+import static java.util.concurrent.CompletableFuture.completedFuture;
+
+public class KafkaPageSink
+        implements ConnectorPageSink
+{
+    private final String topicName;
+    private final List<KafkaColumnHandle> columns;
+    private final RowEncoder keyEncoder;
+    private final RowEncoder messageEncoder;
+    private final ConnectorSession session;
+    private final KafkaProducer<byte[], byte[]> producer;
+
+    public KafkaPageSink(
+            String topicName,
+            List<KafkaColumnHandle> columns,
+            RowEncoder keyEncoder,
+            RowEncoder messageEncoder,
+            ConnectorSession session,
+            PlainTextKafkaProducerFactory producerFactory)
+    {
+        this.topicName = requireNonNull(topicName, "topicName is null");
+        this.columns = (requireNonNull(ImmutableList.copyOf(columns), "columns is null"));
+        this.keyEncoder = requireNonNull(keyEncoder, "keyEncoder is null");
+        this.messageEncoder = requireNonNull(messageEncoder, "messageEncoder is null");
+        this.session = requireNonNull(session, "session is null");
+        this.producer = requireNonNull(producerFactory.create(new ByteArraySerializer(), new ByteArraySerializer()), "producerFactory is null");
+    }
+
+    @Override
+    public CompletableFuture<?> appendPage(Page page)
+    {
+        for (int position = 0; position < page.getPositionCount(); position++) {
+            for (int channel = 0; channel < page.getChannelCount(); channel++) {
+                appendColumn(messageEncoder, page, channel, position);
+            }
+            producer.send(new ProducerRecord<>(topicName, messageEncoder.toByteArray()));
+            keyEncoder.clear();
+            messageEncoder.clear();
+        }
+        producer.flush();
```

Comment:
`producer.send` returns a `Future`, not a `CompletableFuture`

Reply:
is there a way to wake up the sender using the producer?

Reply:
Maybe sth like this would do?
```java
    private static class CountingSendCallback
            implements Callback
    {
        private final AtomicLong remaining;
        private final CompletableFuture<?> future;

        private CountingSendCallback(long pageSize) {
            this.remaining = new AtomicLong(pageSize);
            this.future = new CompletableFuture();
        }

        @Override
        public void onCompletion(RecordMetadata metadata, Exception exception)
        {
            // TODO possibly handle exception here
            remaining.decrementAndGet();
            completeFutureIfDone();
        }

        private void completeFutureIfDone()
        {
            if (remaining.get() <= 0) {
                future.complete(null);
            }
        }

        public CompletableFuture<?> getFuture()
        {
            return future;
        }
    }

    @Override
    public CompletableFuture<?> appendPage(Page page)
    {
        CountingSendCallback callback = new CountingSendCallback(page.getPositionCount());
        for (int position = 0; position < page.getPositionCount(); position++) {
   ....
            producer.send(new ProducerRecord<>(topicName, keyEncoder.toByteArray(), messageEncoder.toByteArray()), callback);
  .....
        }
        return callback.getFuture();
    }
```

Reply:
This also brings my attention to error handling. What would happen in current codebase if writing operation fails? Would that be propagated to INSERT query in presto. Maybe `producer.close();` would fail in such case but I am not sure. Do you know?

Reply:
> is there a way to wake up the sender using the producer?

Good point. Without it the code I pasted will not work. As the sending may basically not happen ever.
So maybe we need explicit `flush()` here or in `finish()`.
Does `producer.close()` we call in `finish()` also flush the sender queue. I skimmed Kafka code and it does not seem so.

It feels the safest for now to get back to what you had originally - flush after each page. And improve in followup PRs as we get more understanding.

Reply:
that looks great, thank you! I think that you're right about handling exceptions in the `onCompletion` callback method. The docs for `KafkaProducer` say that `producer.close()` blocks until all previously sent requests complete.

Reply:
> that looks great

Not that much really :) - unless we can trigger actual sending to happen It is prone to starvation. As I wrote above. I think safest would be to do what you did originally to flush after each page or only in finish.

Reply:
ok, we can come back to this later. Seems like it would still be smart to have a callback with an `onCompletion` method to handle any errors.